net/multi_listen.go

```diff
@@ -0,0 +1,153 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package net
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync"
+	"sync/atomic"
+)
+
+// connErrIndexTriplet packs results of accept call on actual listeners along with the
+// index of the listener. It is used for communication between main and listener goroutines.
+type connErrIndexTriplet struct {
+	index int
+	conn  net.Conn
+	err   error
+}
+
+// multiListener implements net.Listener
+type multiListener struct {
+	latestAcceptedIndex atomic.Int32
+	addrs               []net.Addr
+	listeners           []net.Listener
+
+	wg             sync.WaitGroup
+	connErrIndexCh chan connErrIndexTriplet
+	closed         atomic.Bool
+}
+
+// compile time check to ensure *multiListener implements net.Listener.
+var _ net.Listener = &multiListener{}
+
+// MultiListen returns net.Listener which can listen for and accept
+// TCP connections on multiple addresses.
+func MultiListen(ctx context.Context, addresses []string) (net.Listener, error) {
+	return multiListen(
+		ctx,
+		addresses,
+		net.ResolveTCPAddr,
+		func(ctx context.Context, network, address string) (net.Listener, error) {
+			var lc net.ListenConfig
+			return lc.Listen(ctx, network, address)
+		})
+}
+
+func multiListen(
+	ctx context.Context,
+	addresses []string,
+	resolveTCPAddrFunc func(network, address string) (*net.TCPAddr, error),
+	listenFunc func(ctx context.Context, network, address string) (net.Listener, error),
+) (net.Listener, error) {
+	ml := &multiListener{
+		connErrIndexCh: make(chan connErrIndexTriplet),
+	}
+
+	for _, address := range addresses {
+		addr, err := resolveTCPAddrFunc("tcp", address)
+		if err != nil {
+			// close all listeners
+			_ = ml.Close()
+			return nil, err
+		}
+
+		var network string
+		host, _, err := net.SplitHostPort(addr.String())
+		if err != nil {
+			// close all listeners
+			_ = ml.Close()
+			return nil, err
+		}
+		switch IPFamilyOf(ParseIPSloppy(host)) {
+		case IPv4:
+			network = "tcp4"
+		case IPv6:
+			network = "tcp6"
+		default:
+			// close all listeners
+			_ = ml.Close()
+			return nil, fmt.Errorf("failed to identify ip family of address '%s", addr.String())
+		}
+
+		l, err := listenFunc(ctx, network, addr.String())
+		if err != nil {
+			// close all listeners
+			_ = ml.Close()
+			return nil, err
+		}
+
+		ml.addrs = append(ml.addrs, addr)
+		ml.listeners = append(ml.listeners, l)
+	}
+
+	for i := range ml.listeners {
```

Comment:
Use
```go
for i, l := range ml.listeners {
```
or, if you accept to drop latestAcceptedIndex:
```go
for _, l := range ml.listeners {
```

Reply:
I can't directly simplify to 
```go
for _, l := range ml.listeners {
```
as the variable `l` captured by the gorutine will always point to the last listener. 
https://go.dev/wiki/CommonMistakes

Reply:
Thanks for the pointer. I didn't know. But this should work, shouldn't it?
```go
	for _, l := range ml.listeners {
		ml.wg.Add(1)
		// spawn a go routine for every listener to wait for incoming connection requests
		go func(l net.Listener) {
			defer ml.wg.Done()
			for {
				conn, err := l.Accept()
				if err != nil && ml.closed.Load() {
					return
				}
				ml.connErrIndexCh <- connErrIndexTriplet{conn: conn, err: err}
			}
		}(l)
	}
```
(at least it passes `go test`)